File: HISTORY - CODE/Comatose v1.1/Python2 [CONVERTED]/pyComa/creature.py
from __future__ import absolute_import
from symbol import symbol, arrow
import board

opening = False

def check(x):
    assert x > 0
    boader = board.BOARD_REF.Boarder()
    return boader[x] == symbol.BD or boader[x] == symbol.CE or boader[x] == symbol.C or x > len(boader)

class creature(object):
    def __init__(self, location, dirr): #SYMBOLS board
        self.loc = location
        self.direction = dirr
    def turn(self): #what if goes off boader?
        u''' turn
        * - determines if the position that the creature would move into is an edge (true)
        '''
        for comp, result in [(arrow.U, self.loc - board.WIDTH),
                            (arrow.D, self.loc + board.WIDTH),
                            (arrow.L, self.loc - 1),
                            (arrow.R, self.loc + 1),
                            (None, None)
                            ]:
            if comp == self.direction:
                x = result
        if x and check(x):
            return True
        return False
    def onemove(self):
        u'''onemove
        * - if turn then the direction is changed clockwise (negative angle)
        * - if !turn then changes the loc (location) to position would move into
        '''
        if self.turn():
            for comp, result in [(arrow.U, arrow.R),
                                (arrow.D, arrow.L),
                                (arrow.L, arrow.U),
                                (arrow.R, arrow.D)
                                ]:
                if comp == self.direction:
                    self.direction = result
                    break
        else:
            for comp, result in [(arrow.U, -board.WIDTH),
                                (arrow.D, +board.WIDTH),
                                (arrow.L, -1),
                                (arrow.R, +1)
                                ]:
                if comp == self.direction:
                    self.loc += result
                    break
    def moving(self):
        u''' moving
        * - when loc (location) changes
        * - ask if there is a B (bomb)
        * - if B then move to corner (upper left @ position 0)
        * - only one loc change
        '''
        if self.loc:
            dirr = arrow.BAD
            d = self.direction
            n = 0
            while True: #do-while equivalent
                dirr = self.direction
                self.onemove()
                if dirr != self.direction and d != self.direction:
                    continue
                else:
                    break
    def cbomb(self, crack, crachv, locOFb):
        u''' cbomb
        * - ask if there is a B (bomb)
        * - if B then move to corner (upper left @ position 0)
        '''
        for n, i in enumerate(locOFb):
            global opening
            if self.loc == i:
                v = 0
                if crachv == 1:
                    v = board.WIDTH
                elif crachv == board.WIDTH:
                    v = 1
                if (self.loc == crack + v or self.loc == crack + crachv + v or self.loc == crack - crachv + v
                    or self.loc == crack - v or self.loc == crack + crachv - v or self.loc == crack - crachv - v
                    or self.loc == crack + crachv * 2 or self.loc == crack - crachv * 2):
                    opening = True
                del locOFb[n]
                self.loc = 0
                break
    def bombwall(self, crack, crachv, locOFb):
        tmp = self.loc
        self.cbomb(crack, crachv, locOFb)
        t = tmp - self.loc
        if 0 > t:
            t *= -1
        return t

class creatureR(creature):
    def __init__(self, location, dirr):
        super(creatureR, self).__init__(location, dirr) #super(creature, self)
    def onemove(self):
        if self.turn():
            for comp, result in [(arrow.U, arrow.L),
                                (arrow.D, arrow.R),
                                (arrow.L, arrow.D),
                                (arrow.R, arrow.U)
                                ]:
                if comp == self.direction:
                    self.direction = result
                    break
        else:
            for comp, result in [(arrow.U, -board.WIDTH),
                                (arrow.D, +board.WIDTH),
                                (arrow.L, -1),
                                (arrow.R, +1)
                                ]:
                if comp == self.direction:
                    self.loc += result
                    break

File: src/C++/creature.cpp
#include "creature.h"

creature::creature(whole location, arrow dir, SYMBOLS board)
	: loc(location),
	direction(dir),
	boarder(board)
{}

creatureR::creatureR(whole location, arrow dir, SYMBOLS board)
	: creature(location, dir, board)
{}

creature::~creature()
{}

creatureR::~creatureR()
{}

bool check(const SYMVECT& boader, const MAX& x)
{
	return (boader[x] == symbol::BD || boader[x] == symbol::CE || boader[x] == symbol::C || x > boader.size());
}

/* turn
* - determines if the position that the creature would move into is an edge (true)
*/
bool creature::turn(const MAX& w) //what if goes off boader?
{
	MAX x;
	switch (direction)
	{
	case arrow::U:
		x = loc - w;
		if (check(*boarder, x)) return true;
		return false;
	case arrow::D:
		x = loc + w;
		if (check(*boarder, x)) return true;
		return false;
	case arrow::L:
		x = loc - 1;
		if (check(*boarder, x)) return true;
		return false;
	case arrow::R:
		x = loc + 1;
		if (check(*boarder, x)) return true;
		return false;
	default:
		return false;
	}
}

/*onemove
* - if turn then the direction is changed clockwise (negative angle)
* - if !turn then changes the loc (location) to position would move into
*/
void creature::onemove(CMAX w)
{
	if (turn(w))
	{
		switch (direction)
		{
		case arrow::U:
			direction = arrow::R;
			return;
		case arrow::D:
			direction = arrow::L;
			return;
		case arrow::L:
			direction = arrow::U;
			return;
		case arrow::R:
			direction = arrow::D;
			return;
        case arrow::BAD:
            direction = arrow::U; //WARNING
            return;
		}
	}
	else
	{
		switch (direction)
		{
		case arrow::U:
			loc -= w;
			return;
		case arrow::D:
			loc += w;
			return;
		case arrow::L:
			loc -= 1;
			return;
		case arrow::R:
			loc += 1;
			return;
        case arrow::BAD:
            return; //WARNING
		}
	}
}

void creatureR::onemove(CMAX w)
{
	if (turn(w))
	{
		switch (direction)
		{
		case arrow::U:
			direction = arrow::L;
			return;
		case arrow::D:
			direction = arrow::R;
			return;
		case arrow::L:
			direction = arrow::D;
			return;
		case arrow::R:
			direction = arrow::U;
			return;
        case arrow::BAD:
            direction = arrow::U; //WARNING
            return;
		}
	}
	else
	{
		switch (direction)
		{
		case arrow::U:
			loc -= w;
			return;
		case arrow::D:
			loc += w;
			return;
		case arrow::L:
			loc -= 1;
			return;
		case arrow::R:
			loc += 1;
			return;
        case arrow::BAD:
            return; //WARNING
		}
	}
}

/* moving
* - when loc (location) changes
* - ask if there is a B (bomb)
* - if B then move to corner (upper left @ position 0)
* - only one loc change
*/
void creature::moving(CMAX w)
{
	if (loc != 0)
	{
		arrow dir, d;
		d = direction;
		do
		{
			dir = direction;
			onemove(w);
		} while (dir != direction && d != direction);
	}
}

/* cbomb
* - ask if there is a B (bomb)
* - if B then move to corner (upper left @ position 0)
*/
void creature::cbomb(CMAX w, MAX crack, MAX crachv, PVINT locOFb)
{
	const MAX tot = locOFb->size();
	for (CNT i = 0; i < tot; i++)
	{
		if (loc == locOFb->at(i))
		{
			MAX v = 0;
			if (crachv == 1) v = w;
			else if (crachv == w) v = 1;
			if (loc == crack + v || loc == crack + crachv + v || loc == crack - crachv + v
				|| loc == crack - v || loc == crack + crachv - v || loc == crack - crachv - v
				|| loc == crack + crachv * 2 || loc == crack - crachv * 2)
			{
				finopening();
			}
			locOFb->erase(locOFb->begin() + i);
			loc = 0;
			break;
		}
	}
}

whole creature::bombwall(CMAX w, MAX crack, MAX crachv, PVINT locOFb)
{
	const int tmp = loc;
	cbomb(w, crack, crachv, locOFb);
	int t = tmp - loc;
	if (0 > t) t *= -1;
	return t;
}

/* opening
* - function for telling if crack is open
*/
bool creature::opening() { return *opens; }
bool creature::finopening() { *opens = true; return *opens; }
bool creature::stopening() { *opens = false; return *opens; }
